### ISSUE TITLE
Flexibility for keyboard to decide on Type A, Type B, or both, for fpc cable support

### ISSUE THREAD

**Issue by sadekbaroudi** [sadekbaroudi/vik#4]
I'm have considered this in the past, and I'm considering it again...

Currently, VIK specifies that you must use a Type A fpc cable. The reality is, it doesn't have to... If I were to offer the option to wire the keyboard the way it is today, or inverted to match the module vik connector, then the keyboard designer could decide whether they wanted to use Type A or Type B. This would not break module compatibility, as the modules would always be wired the same...

So, currently there are the following symbols for the schematic:
vik-keyboard-connector
vik-module-connector

Instead, I'd include:
vik-keyboard-connector-type-a
vik-keyboard-connector-type-b
vik-module-connector

The requirement would be that the VIK keyboard certification card would have to include which type it uses. The keyboard distributor/designer/open source repo would have to make sure it clearly specifies this (by means of the VIK certification or otherwise).

Pro: more flexibility for keyboard designs to route the cable much more cleanly if needed
Con: people would need to be aware and carry both types of cables if they planned on using multiple boards with VIK connectors on the keyboard side

**Comment by thomasbaart** [comment]
This issue was just mentioned on the splitkb.com Discord server. Our modules have the VIK pinout reversed, and our VIK port on the keyboard is mounted on the bottom side. As such, our documentation currently (sparsely) mentions:

> Our keyboards are compatible with any other non halcyon VIK module using a standard type A cable (Same side contacts).

and:

> Our modules are compatible with any other VIK keyboard using a type D cable (Opposite side contacts).

It'd probably be useful to have some kind of distinction, like putting some symbol next to both the module and the keyboard port. If both of them match, you use a same-side cable, otherwise you use a different-side cable.

**Comment by sadekbaroudi** [comment]
@thomasbaart - sounds good. I just needed a nudge to implement this one. I can take care of this soon. I'll update the footprints as described in my first comment in this issue, and then update the VIK specs themselves with images to highlight the distinction. I know you implemented it on your board(s). Did you put any markings or anything that you'd like me to repurpose? Otherwise, I'll just make up my own.

**Comment by thomasbaart** [comment]
TLDR: The _only_ reason why you would need to invert the pin sequence is to be able to _neatly tuck away the cable_. Doing so _requires_ the use of a different cable type.

Let's determine what makes a set of a keyboard, module and cable compatible with one another.

# Flat cable types

Flat connector cables can be grouped by three classes:

1. Cables with the pins on the _same_ side, where one side of the pins is shielded by insulation and/or is supported by stiff material.
    - Flat cables types that conform to this description are "Type A", "Type C" and "Type F".
2. Cables with the pins on _opposing_ sides, where one side of the pins is shielded by insulation and/or is supported by stiff material.
    - Flat cables types that conform to this description are "Type B" and "Type D".
3. Cables with one or both sets of pins exposed on _both sides_, without any shielding on one or both connectors.
    - Flat cables types that conform to this description are "Type E", "Type G", "Type H" and "Type I".

The cable type _only_ matters for making a connection with the connector. The connectors I've seen can only connect with the cable on the PCB side, so the side of the cable facing the PCB would need to be exposed. A cable in the third "class", with both sides exposed, could always be connected, though it'd be more fragile. The reason why you want your cable to be shielded on one side is to _prevent_ mixing up the sequence of pins (thus potentially shorting various pins on your controller), as you'll see below.


# Relative orientation of connectors

The readme states the following:

> ... the cable will automatically invert the order when connected.

This is _technically_ not true. It's not the cable doing the inverting, but rather the orientation of the connectors relative to one another:

- The same connector is used on both the keyboard and the module.
- One of the connectors needs to be rotated by 180º to match the other end of the cable.
- The cable is just a flexible piece of material with traces running along it, so if pin 1 is connected on the left side, it will remain on the left. With a rotated connector, that left pin will become pin 12 on the other side.

This distinction matters because the cable in of itself _does not_ invert the sequence of pins. The cable _can invert_ the sequence only if the cable itself is _rotated along its axis_. This can be done by bending it over itself an _uneven_ number of times, such as with a single 45º bend, or a half-turn corkscrew bend.

If the cable is bent perpendicular to itself (folded in half), the pin that's on the "left side" of the cable will still remain on the left side. It does not, in of itself, require the pin assignment to be inverted.

In short, if the orientation of the connectors is the same (both facing the same cardinal direction, and both mounted on the same side of the PCB), then a pin sequence inversion is not needed.


# Compatibility

What determines whether a keyboard is compatible with a module, then, is defined by the relative orientation of its connectors and whether the orientation of the cable is flipped along the way.

Let's go through the same example for the Vulpes Minora, which was supplied as the first example keyboard in the VIK specification:

- The connector is placed on the _top_ of the right side of the keyboard and faces _west_. I assume this uses the pin sequence outlined in the README, so pin 1 is `3.3V`, and pin 12 is `SCLK`.
- We're looking at the keyboard from the top, so in our perspective, pin 1 is on the _north_ side.
- Let's plug in the cable. It's a type A cable, as per the documentation, and as it simply connects to the connector, its north side will be connected to pin 1, same as the keyboard's connector.
- If we then connect a module with its connector being on the _top_ side, that connector would face _east_, and its pin 1 would face _south_: the opposite of what our cable has. It follows that we have to _invert_ the pin sequence.

Easy enough! We made sure both connectors are facing the same side of their respective PCBs, we kept their orientation the same during their installation as well as during the design phase, and so we're able to use a type A cable with the contacts on the same side, just as the specifications intended.

What if we want to make a module that has its connector on its bottom side?

- We're starting out with the same right side of a Vulpes Minora, so the connector is still placed on the _top_ side of the keyboard facing _west_, with pin 1 facing _north_.
- With a type A cable as specified by the documentation, the other end of the cable is still exposed on the bottom side. If we want to connect it to our module's bottom side, we need to _twist the cable along its axis_, thus inverting the direction in which pin 1 faces: pin 1 is now connected to the cable's _south_ side.
- With the module's connector on the _bottom_ side, it still faces _east_. Its pin 1 would face _north_, and we would still have to _invert_ the module's pin mapping.

Twisting the cable along its axis might be a bit of a snag, but not too difficult - we probably want to fold the cable by roughly 45º anyway as the module would be mounted in a 3D printed case in the middle. Nice!

Let's now take a look at an example that I'm familiar with: a Halcyon series keyboard, along with a Halcyon series module:

- The connector is placed on the _bottom_ of the keyboard and faces _south_. It uses the pin sequence as outlined in the README: pin 1 is `3.3V`, and pin 12 is `SCLK`.
- We're looking at our keyboard from the top, so in our perspective, pin 1 is on the _east_ side.
- Let's plug in the flat cable. It's threaded through a slot in the PCB toward the top, where a module is to be installed. Folding the cable over itself does not invert the pin sequence, so pin 1 still faces _east_.
- The module's connector is placed on the _bottom_ of its PCB, and it too faces _south_, because that's where the cable will be coming from. This orientation means that the connector's pin 1 also faces _east_.
- With both pin 1 on the cable as well as pin 1 on the connector both facing _east_, and with our cable _not being twisted_, we are _not inverting_ the sequence of pins on our module.
- If we would use a type A cable per the specifications, then we'd run into an issue now: the cable would need to be inverted in order for the exposed contacts to connect with the connector! If we'd twist the cable, we'd also need to invert the pin sequence. We're now faced with a choice: do we go for a tidy cable and use one with  _contacts on opposing sides_ without inverting the pin order, or do we properly adhere to specifications, use a type A cable, twist it and invert the pin order?

We went with the nice and tidy solution, making for a less awkward install and tidier cable, at the cost of _breaking compatibility with the VIK specifications_.


# Conclusion

From all this, we can conclude the following:

- Twisting the cable along its axis an uneven number of times will invert the pin order.
- Placing the connector on an _opposite side_ will also invert the pin order. Given the same cable as before, it requires you to twist the cable along its axis an uneven number of times.

With that in mind, it can be stated that:

- Given that the pinout on the keyboard is always the same and given that a _type A_ cable with contacts on the same sides is used, you _will invert_ the pin mapping on a module.
- Given that the pinout on the keyboard is always the same and given that a _type B_ cable with contacts on opposing sides is used, you _will not invert_ the pin mapping on a module.

And so:

- Given that the pinout on the keyboard is _always the same_, _the module can decide_ which type of cable the user should use: either _type A_ with contacts on the same sides, or _type B_ with contacts on opposing sides.

I think the most simple solution is to require modules to indicate whether they want a type A or type B cable, where type A would be the current default, and type B would be for modules that wish to not invert the pin mapping.

It could be as simple as appending a character to the text already present next to the connector, such as "3v3 A" and "3v3 B", or "VIK IN A" and "VIK IN B".

**Comment by thomasbaart** [comment]
Or as Harvey eloquently put it, putting my entire essay to shame:

> The insulation on one side of the cable forces its orientation relative to the connector, so it doesn’t matter how many twists or folds you make. Type A will always require an inversion, while type B will always require the sequence to remain the same. The only requirement is that the keyboard will always retain the same sequence, if you start to mess with that too, then it starts to get really complicated.
